rosjava/src/main/java/org/ros/internal/node/topic/DefaultPublisher.java

```diff
@@ -107,9 +109,16 @@ public boolean getLatchMode() {
 
   @Override
   public void shutdown(long timeout, TimeUnit unit) {
-    signalOnShutdown(timeout, unit);
-    outgoingMessageQueue.shutdown();
-    listeners.shutdown();
+    try{
+      shutdownLatch = new CountDownLatch(listeners.size());
+      signalOnShutdown(timeout, unit);
+      shutdownLatch.await(timeout, unit);
+      outgoingMessageQueue.shutdown();
+      listeners.shutdown();
+    }
+    catch (InterruptedException e) {
+      log.error(e.getMessage(), e);
+    }
```

Comment:
Please shrink the `try / catch` block only to the `await` call, and add a comment saying that in the normal workflow, `signalOnMasterUnregistration**` should finish execution before shutting down the listeners.